File: scripts/audit_bank.py
#!/usr/bin/env python3
"""Audit generatedBank.js for broken answers, NaNs, and bad choices."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

BANK = Path(__file__).resolve().parent.parent / "src" / "generatedBank.js"


def extract_problems(text: str) -> list[dict]:
    """Best-effort parse of export const GENERATED_BANK = {...}"""
    # Convert JS object-ish to JSON: quote keys, strip trailing commas carefully.
    m = re.search(r"export const GENERATED_BANK = (\{[\s\S]*\n\});", text)
    if not m:
        raise SystemExit("GENERATED_BANK not found")
    raw = m.group(1)
    # Problems are objects with known keys — walk with a simpler regex split by topic arrays.
    problems = []
    # Match each problem object roughly
    for pm in re.finditer(
        r"\{\s*source:\s*(\"(?:\\.|[^\"])*\")\s*,\s*title:\s*(\"(?:\\.|[^\"])*\")\s*,\s*"
        r"prompt:\s*(\"(?:\\.|[^\"])*\")\s*,\s*finalAnswer:\s*(\"(?:\\.|[^\"])*\")",
        text,
    ):
        start = pm.start()
        # Find matching closing for this object — scan braces from start
        depth = 0
        end = start
        for i, ch in enumerate(text[start:], start):
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0:
                    end = i + 1
                    break
        block = text[start:end]

        def js_str(s: str) -> str:
            return json.loads(s)

        source = js_str(pm.group(1))
        title = js_str(pm.group(2))
        prompt = js_str(pm.group(3))
        final = js_str(pm.group(4))
        choices = []
        for cm in re.finditer(
            r'\{\s*id:\s*"([a-d])"\s*,\s*latex:\s*("(?:\\.|[^"])*")\s*,\s*label:\s*("(?:\\.|[^"])*")\s*\}',
            block,
        ):
            choices.append(
                {
                    "id": cm.group(1),
                    "latex": js_str(cm.group(2)),
                    "label": js_str(cm.group(3)),
                }
            )
        steps = []
        for sm in re.finditer(
            r'\{\s*title:\s*("(?:\\.|[^"])*")\s*,\s*body:\s*("(?:\\.|[^"])*")\s*\}',
            block,
        ):
            steps.append({"title": js_str(sm.group(1)), "body": js_str(sm.group(2))})
        problems.append(
            {
                "source": source,
                "title": title,
                "prompt": prompt,
                "finalAnswer": final,
                "choices": choices,
                "steps": steps,
                "block": block,
            }
        )
    return problems


BAD_PATTERNS = [
    (r"\\text\{NaN\}", "NaN in latex"),
    (r"\bNaN\b", "NaN text"),
    (r"\\text\{Infinity\}", "Infinity latex"),
    (r"\bInfinity\b", "Infinity text"),
    (r"zoo", "sympy complex infinity zoo"),
    (r"\\frac\{\}", "empty frac numerator/denom"),
    (r"undefined", "undefined"),
    (r"2 1\b", "broken pi-strip '2 1'"),
    (r"S = 2 1", "broken pi factor"),
]


def audit(problems: list[dict]) -> list[dict]:
    findings = []
    for i, p in enumerate(problems):
        issues = []
        blob = p["finalAnswer"] + "\n" + "\n".join(
            c["latex"] for c in p["choices"]
        ) + "\n" + "\n".join(s["body"] for s in p["steps"])

        for pat, label in BAD_PATTERNS:
            if re.search(pat, blob):
                issues.append(label)

        # Choice quality
        lats = [c["latex"] for c in p["choices"]]
        labels = [c["label"] for c in p["choices"]]
        if len(p["choices"]) != 4:
            issues.append(f"expected 4 choices, got {len(p['choices'])}")
        if labels.count("Correct") != 1:
            issues.append(f"Correct label count={labels.count('Correct')}")
        # duplicate latex among choices
        if len(set(lats)) < len(lats):
            issues.append("duplicate choice latex")
        # correct choice is NaN
        for c in p["choices"]:
            if c["label"] == "Correct" and ("NaN" in c["latex"] or c["latex"] in ("0", "1", "2") and "NaN" in blob):
                if "NaN" in c["latex"]:
                    issues.append("correct choice is NaN")
        # final answer empty / nan
        if "NaN" in p["finalAnswer"]:
            issues.append("finalAnswer is NaN")
        # step final disagrees with finalAnswer when step has NaN
        for s in p["steps"]:
            if s["title"] == "Final simplified value" and "NaN" in s["body"] and "NaN" not in p["finalAnswer"]:
                issues.append("steps final NaN but finalAnswer ok")

        if issues:
            findings.append(
                {
                    "index": i,
                    "title": p["title"],
                    "prompt": p["prompt"][:100],
                    "source": p["source"][:80],
                    "issues": sorted(set(issues)),
                }
            )
    return findings


def main() -> None:
    text = BANK.read_text(encoding="utf-8")
    problems = extract_problems(text)
    print(f"parsed {len(problems)} problems")
    findings = audit(problems)
    print(f"problems with issues: {len(findings)}")
    # group by issue type
    by = {}
    for f in findings:
        for iss in f["issues"]:
            by.setdefault(iss, []).append(f)
    for iss, items in sorted(by.items(), key=lambda kv: -len(kv[1])):
        print(f"\n## {iss} ({len(items)})")
        for it in items[:12]:
            print(f"  - [{it['title']}] {it['prompt']}")
        if len(items) > 12:
            print(f"  ... +{len(items)-12} more")
    # write report
    out = Path(__file__).resolve().parent.parent / "agent-tools" / "bank-audit.json"
    out.parent.mkdir(exist_ok=True)
    out.write_text(json.dumps(findings, indent=2), encoding="utf-8")
    print(f"\nwrote {out}")
    sys.exit(1 if findings else 0)


if __name__ == "__main__":
    main()

File: scripts/audit_bank.mjs
/** Audit GENERATED_BANK for NaN, broken choices, and bad latex. */
import { writeFileSync } from "fs";
import katex from "katex";

// Windows path fix
const { GENERATED_BANK } = await import("../src/generatedBank.js");

const BAD = [
  [/\\text\{NaN\}|\bNaN\b/, "NaN"],
  [/\\text\{Infinity\}|\bInfinity\b|\bzoo\b/, "Infinity/zoo"],
  [/\b2 1\b/, "broken pi-strip '2 1'"],
  [/\b1 1\b/, "broken pi-strip '1 1'"],
];

/** Match main.js: collapse accidental double-backslashes before KaTeX. */
const normalizeLatex = (value) => String(value).replace(/\\\\/g, "\\");

function bracesBalanced(s) {
  let depth = 0;
  for (const ch of s) {
    if (ch === "{") depth++;
    else if (ch === "}") {
      depth--;
      if (depth < 0) return false;
    }
  }
  return depth === 0;
}

function katexRenders(latex) {
  try {
    const html = katex.renderToString(normalizeLatex(latex), {
      throwOnError: true,
      strict: "ignore",
      trust: false,
    });
    return !html.includes("katex-error");
  } catch {
    return false;
  }
}

const findings = [];
let total = 0;

for (const [topic, problems] of Object.entries(GENERATED_BANK)) {
  problems.forEach((p, idx) => {
    total++;
    const issues = [];
    const choices = p.choices || [];
    const steps = p.steps || [];
    const blob = [
      p.finalAnswer || "",
      ...choices.map((c) => c.latex || ""),
      ...steps.map((s) => `${s.title}\n${s.body}`),
    ].join("\n");

    for (const [re, label] of BAD) {
      if (re.test(blob)) issues.push(label);
    }

    if (choices.length !== 4) issues.push(`choices count=${choices.length}`);
    const correct = choices.filter((c) => c.label === "Correct");
    if (correct.length !== 1) issues.push(`Correct count=${correct.length}`);

    const lats = choices.map((c) => c.latex);
    if (new Set(lats).size < lats.length) issues.push("duplicate choice latex");

    // Correct choice should not be trivial garbage when answer is long
    if (correct[0] && /NaN/.test(correct[0].latex || "")) issues.push("correct is NaN");

    // Step final should not be NaN if finalAnswer is fine
    const finalStep = steps.find((s) => /final simplified/i.test(s.title || ""));
    if (finalStep && /NaN/.test(finalStep.body || "") && !/NaN/.test(p.finalAnswer || "")) {
      issues.push("step final NaN but finalAnswer ok");
    }

    for (const c of choices) {
      const lx = c.latex || "";
      if (!bracesBalanced(lx)) issues.push("unbalanced braces in choice");
      // Incomplete \frac{num} with no denominator — classic pi-strip debris
      if (/\\frac\s*\{[^{}]*\}\s*$/.test(lx) || /\\frac\s*\{[^{}]*\}(?!\s*\{)/.test(lx)) {
        // second pattern is noisy for nested fracs; rely on KaTeX below
      }
      if (!katexRenders(lx)) issues.push(`katex fail choice ${c.id}`);
    }

    if (issues.length) {
      findings.push({
        topic,
        idx,
        title: p.title,
        prompt: (p.prompt || "").slice(0, 120),
        issues: [...new Set(issues)],
      });
    }
  });
}

// Summarize
const by = {};
for (const f of findings) {
  for (const iss of f.issues) {
    (by[iss] ||= []).push(f);
  }
}

console.log(`audited ${total} problems`);
console.log(`with issues: ${findings.length}`);
for (const [iss, items] of Object.entries(by).sort((a, b) => b[1].length - a[1].length)) {
  console.log(`\n## ${iss} (${items.length})`);
  for (const it of items.slice(0, 8)) {
    console.log(`  - [${it.topic}#${it.idx}] ${it.prompt}`);
  }
  if (items.length > 8) console.log(`  ... +${items.length - 8} more`);
}

writeFileSync(
  new URL("../agent-tools/bank-audit.json", import.meta.url),
  JSON.stringify(findings, null, 2)
);

// Spotlight the reported surface problem
const surf = GENERATED_BANK.surface || [];
const hit = surf.find((p) => /x\^\{3\/2\}/.test(p.prompt) && /\[0,1\]/.test(p.prompt));
if (hit) {
  console.log("\n=== spotlight y=(2/3)x^{3/2} on [0,1] ===");
  console.log("finalAnswer:", hit.finalAnswer);
  console.log("choices:", hit.choices.map((c) => c.latex));
  const lower = hit.steps.find((s) => /lower bound/i.test(s.title));
  const mult = hit.steps.find((s) => /constant factor/i.test(s.title));
  const fin = hit.steps.find((s) => /final simplified/i.test(s.title));
  console.log("lower:", lower?.body);
  console.log("mult:", mult?.body);
  console.log("fin:", fin?.body);
}

process.exit(findings.length ? 1 : 0);
